src/lib/api/index.ts

```diff
@@ -101,78 +103,87 @@ export async function post(
     langs = opts.langs.slice(0, 3)
   }
 
-  let res
-  try {
-    opts.onStateChange?.(t`Posting...`)
-    res = await agent.post({
+  const rkey = TID.nextStr()
+  const uri = `at://${agent.assertDid}/app.bsky.feed.post/${rkey}`
+  const date = new Date().toISOString()
+
+  const writes: ComAtprotoRepoApplyWrites.Create[] = []
+
+  // Create post record
+  {
+    const record: AppBskyFeedPost.Record = {
+      createdAt: date,
       text: rt.text,
       facets: rt.facets,
       reply,
       embed,
       langs,
       labels,
-    })
-  } catch (e: any) {
-    logger.error(`Failed to create post`, {
-      safeMessage: e.message,
-    })
-    if (isNetworkError(e)) {
-      throw new Error(
-        t`Post failed to upload. Please check your Internet connection and try again.`,
-      )
-    } else {
-      throw e
     }
+
+    writes.push({
+      $type: 'com.atproto.repo.applyWrites#create',
+      collection: 'app.bsky.feed.post',
+      rkey: rkey,
+      value: record,
+    })
   }
 
+  // Create threadgate record
   if (draft.threadgate.some(tg => tg.type !== 'everybody')) {
-    try {
-      // TODO: this needs to be batch-created with the post!
-      await writeThreadgateRecord({
-        agent,
-        postUri: res.uri,
-        threadgate: createThreadgateRecord({
-          post: res.uri,
-          allow: threadgateAllowUISettingToAllowRecordValue(draft.threadgate),
-        }),
-      })
-    } catch (e: any) {
-      logger.error(`Failed to create threadgate`, {
-        context: 'composer',
-        safeMessage: e.message,
-      })
-      throw new Error(
-        t`Failed to save post interaction settings. Your post was created but users may be able to interact with it.`,
-      )
-    }
+    const record = createThreadgateRecord({
+      createdAt: date,
+      post: uri,
+      allow: threadgateAllowUISettingToAllowRecordValue(draft.threadgate),
+    })
+
+    writes.push({
+      $type: 'com.atproto.repo.applyWrites#create',
+      collection: 'app.bsky.feed.threadgate',
+      rkey: rkey,
+      value: record,
+    })
   }
 
+  // Create postgate record
   if (
     draft.postgate.embeddingRules?.length ||
     draft.postgate.detachedEmbeddingUris?.length
   ) {
-    try {
-      // TODO: this needs to be batch-created with the post!
-      await writePostgateRecord({
-        agent,
-        postUri: res.uri,
-        postgate: {
-          ...draft.postgate,
-          post: res.uri,
-        },
-      })
-    } catch (e: any) {
-      logger.error(`Failed to create postgate`, {
-        context: 'composer',
-        safeMessage: e.message,
-      })
+    const record: AppBskyFeedPostgate.Record = {
+      ...draft.postgate,
+      createdAt: date,
+      post: uri,
+    }
```

Comment:
Similar here— this should work as-is but I'd recommend setting `$type` on the record anyway.